### PR TITLE
Mark mirror down if replication keeps crash.

### PR DESCRIPTION
**Pull request by JunfengYang**
For GPDB FTS, if the primary, mirror replication keeps crash
continuously and attempt to crate replication connection too many times,
FTS should mark the mirror down. Otherwise, it may block other processes.
If the connection established, clear the attempt count to 0.

The solution for this is:
1. Use replication slot to track current continuously failure times for its primary-mirror replication.
    - the count gets reset once the replication starts streaming, or restarted.

2. When handling each probe query, `GetMirrorStatus` will check the current mirror status and the failure count from its replication slot. If the count exceeds the limit, the retry test will ignore the last replication connection time since it gets refreshed when new walsender starts.(since in current case, the walsender keeps restart.)

3. On FTS bgworker. If mirror down and retry set to false, mark the mirror down.

From the replication slot, it has two main usages:
1. For logical replication. https://www.postgresql.org/docs/current/logical-replication.html
2. Replication slots provide an automated way to ensure that the master does not remove WAL segments until they have been received by all standbys. https://www.postgresql.org/docs/current/warm-standby.html#STREAMING-REPLICATION-SLOTS

So, use the replication slot to track replication failure times between standby seems not so bad?

We can not record the continuously failure for replication in FTS is because FTS only retrieves the mirror status from walsender. But walsender get restarted after a failure. There's no way to count the failure if the FTS sleeps and waits for the next probe. Since during the FTS sleep period, replication may be normal, start streaming, and crash again when the next FTS probe happens.

## Here are some reminders before you submit the pull request
- [x] Add tests for the change
- [ ] Document changes
- [ ] Communicate in the mailing list if needed
- [x] Pass `make installcheck`
- [ ] Review a PR in return to support the community

### CODE REVIEW

**Comment by JunfengYang** [comment]
The travis-ci/pr already passed, I don't know why the status not updated.

File: src/backend/cdb/cdbvars.c [hunk 1]
```diff
@@ -135,6 +135,12 @@ int			gp_fts_probe_interval = 60;
  */
 int gp_fts_mark_mirror_down_grace_period = 30;
 
+/*
+ * If primary-mirror replication attempts to connect continuously and exceed
+ * this count, mark the mirror down dirextly.
```
Comment (ashwinstar) [inline src/backend/cdb/cdbvars.c:140]: s/dirextly/directly (ideally I think don't need "directly", mark mirror down if good)

File: src/backend/replication/slot.c [hunk 2]
```diff
@@ -568,6 +569,58 @@ ReplicationSlotMarkDirty(void)
 	SpinLockRelease(&slot->mutex);
 }
 
+/*
+ * For GPDB FTS purpose, retrieve how many replication connection attempts
+ * to crate replication connection.
```
Comment (ashwinstar) [inline src/backend/replication/slot.c:574]: s/crate/create

File: src/include/replication/slot.h [hunk 3]
```diff
@@ -96,6 +96,14 @@ typedef struct ReplicationSlot
 	bool		just_dirtied;
 	bool		dirty;
 
+	/*
+	 * For GPDB FTS purpose, if the the primary, mirror replication keeps crash
+	 * continuously and attempt to crate replication connection too many times,
```
Comment (ashwinstar) [inline src/include/replication/slot.h:101]: s/crate/create

**Comment by JunfengYang** [comment]
@ashwinstar 
> It would be helpful and better to have the approach explained in commit message and PR description. Reviewers have to dig through code to find the solution. Commit message should describe problem and solution both.

Big thanks for your suggestion!

> Now on solution I dislike modifying replication slot on-disk representation to persist this information. First replication slot has no connection with this info/functionality. Second its upstream and this unnecessarily modifies upstream behavior. Users should be able to no use/completely disable replication slot but still use replication and hence this error handling piece should not be connected with replication slot at all.

Actually, the disk representation for the replication slot is not changed, the `con_attempt_count` always fill with 0 once read from disk. And the new added filed is limited to primary-mirror replication.
From the replication slot, it has two main usages:
1. For logical replication.
2. Replication slots provide an automated way to ensure that the master does not remove WAL segments until they have been received by all standbys.
So, use the replication slot to track replication failure times between standby seems not so bad?

> Is there any possibility of handling this on FTS side instead of primary side (I haven't looked at code)? Like FTS receives retry request since mirror is not up yet, can we code logic there to retry only N times and after that instead of retrying mark mirror down. Given FTS is long running process, it should be able to maintain the state more easily compared to primary needing to persist this info on-disk.

For your solution,
the FTS fork a new process to deal with the probe message. It only returns the current primary, mirror status which gets from walsender. It seems not suitable to record the previous probe message's result.
And for the FTS bgwork, here's 1 min interval for each probe, and each probe retries 5 times(default value) and then give up, clean all context. But the primary-mirror replication may only crash a few times, not reach the appropriate limit within the 5 times retry.

Well, the current solution is handled on the FTS side, the replication slot only tracks the failure times for the replication. And the FTS retrieve the failure times and then decides which mirror to mark as down.

File: src/backend/cdb/cdbvars.c (hunk 1 above)
Comment (JunfengYang) [reply src/backend/cdb/cdbvars.c:140]: haha, thank a lot!

File: src/backend/replication/slot.c (hunk 2 above)
Comment (JunfengYang) [reply src/backend/replication/slot.c:574]: Nice catch!

File: src/include/replication/slot.h (hunk 3 above)
Comment (JunfengYang) [reply src/include/replication/slot.h:101]: Bit thanks!

**Comment by ashwinstar** [comment]
> Actually, the disk representation for the replication slot is not changed, the con_attempt_count always fill with 0 once RestoreSlotFromDisk called. And the new added filed is limited to primary-mirror replication.
From the replication slot, it has two main usages:

For logical replication.
Replication slots provide an automated way to ensure that the master does not remove WAL segments until they have been received by all standbys.
So, use the replication slot to track replication failure times between standby seems not so bad?

**Comment by ashwinstar** [comment]
Sorry by mistake closed it. Reopened it now.

> Actually, the disk representation for the replication slot is not changed, the `con_attempt_count` always fill with 0 once `RestoreSlotFromDisk` called. And the new added filed is limited to primary-mirror replication.
> From the replication slot, it has two main usages:
> 
> 1. For logical replication.
> 2. Replication slots provide an automated way to ensure that the master does not remove WAL segments until they have been received by all standbys.
> 
> So, use the replication slot to track replication failure times between standby seems not so bad?

Thanks for clarifying on-disk structure is not changed. Missed going through full PR. Now, I have taken complete look at the changes. Still feel its little odd to overload replication slot structure with maintaining this information. `replica_disconnected_at` and `con_attempt_count` maintained in same structure seems more logical, as those are connected properties and used for same purpose. 

What are your thoughts on maintaining `con_attempt_count` alongside `replica_disconnected_at` instead of ReplicationSlot structure? When updating `replica_disconnected_at`  bump the `con_attempt_count`. On checking `con_attempt_count` and reporting mirror down instead retry, reset the same.

**Comment by JunfengYang** [comment]
@ashwinstar 
> What are your thoughts on maintaining `con_attempt_count` alongside `replica_disconnected_at` instead of ReplicationSlot structure? When updating `replica_disconnected_at` bump the `con_attempt_count`. On checking `con_attempt_count` and reporting mirror down instead retry, reset the same.

Actually, our code about the primary-mirror replication has an assumption which I disagree, that is we will only have one replication pair. But in the future we may support more than one mirror I think.

At first, I was trying to use `WalSnd` to count the failure times. The `WalSnd` slot is from `WalSndCtl` which in shared memory. For each walsender process startup, it'll init a `WalSnd` slot. And free it once the process exit. So when the FTS probe happens, the `WalSnd` may already get freed.

In our case, the walsender keeps restart and acquire a new `WalSnd` slot each time. If we only have one replication, it's ok to check the freed `WalSnd` slot during FTS since we always use the same slot, although the code is hacky. But once we enable multi-mirrors, the `WalSnd` slot we get may not belong to the same primary-mirror, so this will not work. The check under `GetMirrorStatus` will mess up.

Only the replication slot represents the pair relation between primary and mirror. If we have primary-mirror1, primary-mirror2. It'll have two replication slots represent them.

**Comment by JunfengYang** [comment]
Although my current implementation is also hardcode for one mirror assumption. But it's should be easy to migrate to multi-mirror implementation.

**Comment by ashwinstar** [comment]
I understand and agree with your concerns @JunfengYang about multi-mirror. Lot more needs to be changed in FTS to support multi-mirror not just this piece about `GetMirrorStatus()` anyway. We don't even know how it would look like. Seems we are moving ahead of ourselves. Even for this approach to work in multi-mirror setup, we will have to create mapping to find which primary-mirror pair is using which slot from FTS side to consult correct slot for `con_attempt_count`.

The main outcome I am looking for is `replica_disconnected_at` and `con_attempt_count` serve the same purpose and hence should stay together and not at different structures/places.

Replication slot is not used only for Streaming replication, it can be used just for archiving in which case using `pg_receivewal`, one could just keep fetching the WAL. Or just pg_basebackup. Plus, assuming the replication slot will always be present for streaming replication pair in-case of multi-mirror might be inaccurate. What if user prefers not to use slot and always full recover mirror. In that case unintentionally will loose out on this FTS enhancement and face same issue this PR is trying to solve for cluster. (I know we currently enforce creating replication slot and have not provided option to opt out of it but when it comes to multi-mirror who knows what option we will provide.) Hence, I feel the intended functionalities are different, so should not be mixed together.

This is just my $0.02 cents. Would appreciate more opinions from Team and I am fine to go with it in-case different from above.

**Comment by JunfengYang** [comment]
@ashwinstar Thanks for your opinion! It helps a lot.
I just think it's weird to check the mirror status through a freed `WalSnd` slot.
How about refactoring both `replica_disconnected_at` and `con_attempt_count` into a new structure.
For example, a shared memory hashtable, and the key is the replication application name which should exist as required. The entry will record the FTS related info.

**Comment by paul-guo-** [comment]
> @ashwinstar Thanks for your opinion! It helps a lot.
> I just think it's weird to check the mirror status through a freed `WalSnd` slot.
> How about refactoring both `replica_disconnected_at` and `con_attempt_count` into a new structure for GPDB only.
> I'll try this solution and open a new PR.

How about logging recent wal sender initilization timestamp in shared memory and calculate if flapping happens or not and send back the result to fts when fts probe happens. e.g.

walsnd_init_timestamp[N];
int next_idx;

When wal sender initialization happens, write timestamp to walsnd_init_timestamp[next_idx]; next_idx = (next_idx+1)%N;

when fts probe happens, send them the flapping result to master fts  process to determine if mark the mirror down or not.

A simple algorithm to judge flapping happens or not is to count the avg. initilization times in a period.

**Comment by JunfengYang** [comment]
> How about logging recent wal sender initilization timestamp in shared memory and calculate if flapping happens or not and send back the result to fts when fts probe happens. e.g.
> 
> walsnd_init_timestamp[N];
> int next_idx;
> 
> When wal sender initialization happens, write timestamp to walsnd_init_timestamp[next_idx]; next_idx = (next_idx+1)%N;
> 
> when fts probe happens, send them the flapping result to master fts process to determine if mark the mirror down or not.
> 
> A simple algorithm to judge flapping happens or not is to count the avg. initilization times in a period.

We only mark the mirror down if the replication keeps crash. Which may hang other queries.
If encounter network issue in a short time and the replication could success in between, which would not hang queries.

And the time interval between each walsender initialization depends on the cluster environment.
Maybe 20 times crash still takes a long avg time. Use avg may not represent the walsender failures.

**Comment by JunfengYang** [comment]
Close this cause a new PR is opened.

**Pull request closed without merging**